uaclient/system/cpu_type.py

```diff
@@ -0,0 +1,179 @@
+        )
+
+
+class PPCCPUTypeCollector(BaseCPUTypeCollector):
+    CPU_TYPE_FILE = "/proc/cpuinfo"
+    CPU_TYPE_REGEX_PATTERN = r"{}\s+:\s+(?P<name>.+)"
+
+    def collect(self):
+        cpu_info = system.load_file(self.CPU_TYPE_FILE)
+        return self._get_field_value(
+            cpu_info=cpu_info,
+            re_pattern=self.CPU_TYPE_REGEX_PATTERN.format("cpu"),
+        )
+
+
+class S390CPUTypeCollector(BaseCPUTypeCollector):
+    CPU_TYPE_FILE = "/proc/sysinfo"
+    CPU_TYPE_REGEX_PATTERN = r"{}:\s+(?P<name>.+)"
+
+    def collect(self):
+        cpu_info = system.load_file(self.CPU_TYPE_FILE)
+        cpu_type = (
+            self._get_field_value(
+                cpu_info=cpu_info,
+                re_pattern=self.CPU_TYPE_REGEX_PATTERN.format("Type"),
+            )
+            or ""
+        )
+        cpu_model = (
+            self._get_field_value(
+                cpu_info=cpu_info,
+                re_pattern=self.CPU_TYPE_REGEX_PATTERN.format("Model"),
+            )
+            or ""
+        )
+
+        return "Type:{}; Model:{}".format(cpu_type, cpu_model)
+
+
+class ArmCPUTypeCollector(BaseCPUTypeCollector):
+    CPU_TYPE_FILE = "/sys/firmware/devicetree/base/model"
+    CPU_FALLBACK_FILE = "/proc/cpuinfo"
+    CPU_TYPE_REGEX_PATTERN = r"{}\s+:\s*(?P<name>.+)"
+
+    def collect(self):
+        try:
+            return system.load_file(self.CPU_TYPE_FILE)
+        except FileNotFoundError:
+            pass
+
+        # start backup plan if devicetree file not present
+        fallback_cpu_info = system.load_file(self.CPU_FALLBACK_FILE)
+        return self._get_fallback_name(
+            cpu_info=fallback_cpu_info,
+            fallback_fields=(
+                "CPU implementer",
+                "CPU architecture",
+                "CPU variant",
+                "CPU part",
+                "CPU revision",
+            ),
+            re_pattern=self.CPU_TYPE_REGEX_PATTERN,
+        )
+
+
+def is_intel_or_amd_processor(processor_type: str):
+    return (
+        processor_type == "amd64"
+        or processor_type == "i386"
+        or processor_type == "i686"
+    )
+
+
+def is_ppc_processor(processor_type: str):
+    return processor_type.startswith("ppc")
+
+
+def is_s390_processor(processor_type: str):
+    return processor_type.startswith("s390")
+
+
+def is_arm_processor(processor_type: str):
+    return processor_type.startswith("arm")
+
+
+def get_cpu_type_collector(processor_type):
+    if is_intel_or_amd_processor(processor_type):
+        return X86CPUTypeCollector()
+    elif is_ppc_processor(processor_type):
+        return PPCCPUTypeCollector()
+    elif is_s390_processor(processor_type):
+        return S390CPUTypeCollector()
+    elif is_arm_processor(processor_type):
+        return ArmCPUTypeCollector()
+    else:
+        raise exceptions.UnknownProcessorType(processor_type=processor_type)
+
+
+def get_cpu_type():
+    processor_type = system.get_dpkg_arch()
```

Comment:
dpkg arch is not necessarily the arch of the cpu.
It would be more technically accurate to use `os.uname().machine` I think

Reply:
@cpaelzer just double checking with you if that is a preferred option, since we had that `uname` discussion in the past

Reply:
We discussed out of band and decided dpkg arch is okay